statev2/src/storage/db.rs

```diff
@@ -195,12 +287,177 @@ mod test {
     }
+
+    /// Tests a read only tx to a table
+    #[test]
+    fn test_ro_tx_simple() {
+        // Create a table and add a value
+        let db = mock_db();
+        db.create_table(TABLE_NAME).unwrap();
+
+        let key = "test_key".to_string();
+        let value = TestValue::dummy();
+
+        db.write(TABLE_NAME, &key, &value).unwrap();
+
+        // Create a read only transaction and read the value twice
+        let tx = db.new_read_tx().unwrap();
+        let v1: Option<TestValue> = tx.read(TABLE_NAME, &key).unwrap();
+        let v2: Option<TestValue> = tx.read(TABLE_NAME, &key).unwrap();
+        tx.commit().unwrap();
+
+        assert_eq!(v1.unwrap(), value);
+        assert_eq!(v2.unwrap(), value);
+    }
+
+    /// Tests a read-write tx to a table
+    #[test]
+    fn test_rw_tx_simple() {
+        // Create a mock DB
+        let db = mock_db();
+        db.create_table(TABLE_NAME).unwrap();
+
+        // Write two keys to the table
+        let key1 = "test_key".to_string();
+        let key2 = "test_key2".to_string();
+        let value1 = TestValue::dummy();
+        let value2 = TestValue {
+            a: 5,
+            ..Default::default()
+        };
+
+        let tx = db.new_write_tx().unwrap();
+        tx.write(TABLE_NAME, &key1, &value1).unwrap();
+        tx.write(TABLE_NAME, &key2, &value2).unwrap();
+        tx.commit().unwrap();
+
+        // Read the values back
+        let tx = db.new_read_tx().unwrap();
+        let v1: Option<TestValue> = tx.read(TABLE_NAME, &key1).unwrap();
+        let v2: Option<TestValue> = tx.read(TABLE_NAME, &key2).unwrap();
+        tx.commit().unwrap();
+
+        assert_eq!(v1.unwrap(), value1);
+        assert_eq!(v2.unwrap(), value2);
+    }
+
+    /// Tests reading and writing to multiple tables in a tx
+    #[test]
+    fn test_multi_table_rw() {
+        // Create the first table
+        let db = mock_db();
+        db.create_table(TABLE_NAME).unwrap();
+
+        let key1 = "test_key".to_string();
+        let key2 = "test_key2".to_string();
+        let value1 = TestValue::dummy();
+        let value2 = TestValue {
+            a: 5,
+            ..Default::default()
+        };
+
+        // Create a second table in the write tx
+        const TABLE_NAME2: &str = "test_table2";
+        let tx = db.new_write_tx().unwrap();
+        tx.create_table(TABLE_NAME2).unwrap();
+
+        tx.write(TABLE_NAME, &key1, &value1).unwrap();
+        tx.write(TABLE_NAME2, &key2, &value2).unwrap();
+        tx.commit().unwrap();
+
+        // Read the values back
+        let tx = db.new_read_tx().unwrap();
+        let v1: Option<TestValue> = tx.read(TABLE_NAME, &key1).unwrap();
+        let v2: Option<TestValue> = tx.read(TABLE_NAME2, &key2).unwrap();
+        tx.commit().unwrap();
+
+        assert_eq!(v1.unwrap(), value1);
+        assert_eq!(v2.unwrap(), value2);
+    }
+
+    /// Tests concurrent transaction isolation
+    #[test]
+    fn test_concurrent_updates() {
```

Comment:
gud test, should we run it multiple times or w/ more threads to have higher confidence that some lock contention occurred?

Reply:
I prefer to keep it simple, the test will be run many times anyways over the course of its lifetime.

Plus, practically speaking, `mdbx` has a global write lock, so there isn't *actually* any chance that these two transactions run concurrently